File: Data Structures Intro/Binary Search Tree/binary-tree-traversal-task-net/BinaryTreeTraversal/TreeNode.cs
using System;
using System.Collections.Generic;

namespace BinaryTreeTraversal
{
    /// <summary>
    /// Represents a node in a binary tree.
    /// </summary>
    /// <typeparam name="T">The type of data stored in the node.</typeparam>
    public class TreeNode<T>
    {
        private T element;
        private TreeNode<T>? left;
        private TreeNode<T>? right;

        /// <summary>
        /// Initializes a new instance of the <see cref="TreeNode{T}"/> class with a value and optional children.
        /// </summary>
        /// <param name="element">The value to store in this node.</param>
        /// <param name="left">The left child node (or null).</param>
        /// <param name="right">The right child node (or null).</param>
        public TreeNode(T element, TreeNode<T>? left, TreeNode<T>? right)
        {
            this.element = element;
            this.left = left;
            this.right = right;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="TreeNode{T}"/> class with a value.
        /// </summary>
        /// <param name="element">The value to store in this node.</param>
        public TreeNode(T element) : this(element, null, null) { }

        /// <summary>
        /// Initializes a new instance of the <see cref="TreeNode{T}"/> class with a value and left child.
        /// </summary>
        /// <param name="element">The value to store in this node.</param>
        /// <param name="left">The left child node (or null).</param>
        public TreeNode(T element, TreeNode<T>? left) : this(element, left, null) { }

        /// <summary>
        /// Gets the data element stored in this node.
        /// </summary>
        public T Element => element;

        /// <summary>
        /// Gets the left child node.
        /// </summary>
        public TreeNode<T>? Left => left;

        /// <summary>
        /// Gets the right child node.
        /// </summary>
        public TreeNode<T>? Right => right;

        /// <summary>
        /// Checks whether this tree is structurally and semantically identical to another tree.
        /// </summary>
        /// <param name="treeNode">Root node of the second given tree.</param>
        /// <returns>True if the trees are identical; otherwise false.</returns>
        public bool Check(TreeNode<T>? treeNode)
        {
            if (treeNode == null) return false; // comparing non-null (this) to null -> false

            if (!EqualityComparer<T>.Default.Equals(this.element, treeNode.element))
                return false;

            bool leftEqual = (this.left == null && treeNode.left == null)
                             || (this.left != null && this.left.Check(treeNode.left));
            bool rightEqual = (this.right == null && treeNode.right == null)
                              || (this.right != null && this.right.Check(treeNode.right));

            return leftEqual && rightEqual;
        }

        /// <summary>
        /// Returns elements of this binary tree using inorder traversal (Left, Root, Right).
        /// </summary>
        /// <returns>List with elements in inorder order.</returns>
        public List<T> GetInorderTraversal()
        {
            var result = new List<T>();
            void Traverse(TreeNode<T>? node)
            {
                if (node == null) return;
                Traverse(node.left);
                result.Add(node.element);
                Traverse(node.right);
            }
            Traverse(this);
            return result;
        }

        /// <summary>
        /// Returns elements of this binary tree using postorder traversal (Left, Right, Root).
        /// </summary>
        /// <returns>List with elements in postorder order.</returns>
        public List<T> GetPostorderTraversal()
        {
            var result = new List<T>();
            void Traverse(TreeNode<T>? node)
            {
                if (node == null) return;
                Traverse(node.left);
                Traverse(node.right);
                result.Add(node.element);
            }
            Traverse(this);
            return result;
        }

        /// <summary>
        /// Returns elements of this binary tree using preorder traversal (Root, Left, Right).
        /// </summary>
        /// <returns>List with elements in preorder order.</returns>
        public List<T> GetPreorderTraversal()
        {
            var result = new List<T>();
            void Traverse(TreeNode<T>? node)
            {
                if (node == null) return;
                result.Add(node.element);
                Traverse(node.left);
                Traverse(node.right);
            }
            Traverse(this);
            return result;
        }

        /// <summary>
        /// Returns elements of this binary tree using level-order traversal (breadth-first).
        /// </summary>
        /// <returns>List with elements in level-order.</returns>
        public List<T> GetLevelOrderTraversal()
        {
            var result = new List<T>();
            var queue = new Queue<TreeNode<T>>();
            queue.Enqueue(this);

            while (queue.Count > 0)
            {
                var current = queue.Dequeue();
                result.Add(current.element);
                if (current.left != null) queue.Enqueue(current.left);
                if (current.right != null) queue.Enqueue(current.right);
            }

            return result;
        }
    }
}

File: Data Structures Intro/Binary Search Tree/Java/binary-tree-traversal-task-java/src/main/java/com/epam/bsp/binary/tree/TreeNode.java
package com.epam.bsp.binary.tree;

import java.util.*;

public class TreeNode<E> {

    private E element;
    private TreeNode<E> left;
    private TreeNode<E> right;

    public TreeNode(E element, TreeNode<E> left, TreeNode<E> right) {
        this.element = element;
        this.left = left;
        this.right = right;
    }

    public TreeNode(E element) {
        this(element, null, null);
    }

    public TreeNode(E element, TreeNode<E> left) {
        this(element, left, null);
    }

    public E getElement() {
        return element;
    }

    public TreeNode<E> getLeft() {
        return left;
    }

    public TreeNode<E> getRight() {
        return right;
    }

    /**
     * Checks whether two given binary trees are the same.
     */
    public boolean check(TreeNode<E> treeNode) {
        if (treeNode == null) return false;
        return isSame(this, treeNode);
    }

    private boolean isSame(TreeNode<E> t1, TreeNode<E> t2) {
        if (t1 == null && t2 == null) return true;
        if (t1 == null || t2 == null) return false;
        if (!(Objects.equals(t1.element, t2.element))) return false;
        return isSame(t1.left, t2.left) && isSame(t1.right, t2.right);
    }

    /**
     * Inorder traversal: left → root → right
     */
    public List<E> getInorderTraversal() {
        List<E> res = new ArrayList<>();
        inorder(this, res);
        return res;
    }

    private void inorder(TreeNode<E> node, List<E> res) {
        if (node == null) return;
        inorder(node.left, res);
        res.add(node.element);
        inorder(node.right, res);
    }

    /**
     * Postorder traversal: left → right → root
     */
    public List<E> getPostorderTraversal() {
        List<E> res = new ArrayList<>();
        postorder(this, res);
        return res;
    }

    private void postorder(TreeNode<E> node, List<E> res) {
        if (node == null) return;
        postorder(node.left, res);
        postorder(node.right, res);
        res.add(node.element);
    }

    /**
     * Preorder traversal: root → left → right
     */
    public List<E> getPreorderTraversal() {
        List<E> res = new ArrayList<>();
        preorder(this, res);
        return res;
    }

    private void preorder(TreeNode<E> node, List<E> res) {
        if (node == null) return;
        res.add(node.element);
        preorder(node.left, res);
        preorder(node.right, res);
    }

    /**
     * Level order traversal (BFS)
     */
    public List<E> getLevelOrderTraversal() {
        List<E> res = new ArrayList<>();
        if (this == null) return res;

        Queue<TreeNode<E>> queue = new LinkedList<>();
        queue.add(this);

        while (!queue.isEmpty()) {
            TreeNode<E> node = queue.poll();
            if (node == null) continue;
            res.add(node.element);
            if (node.left != null) queue.add(node.left);
            if (node.right != null) queue.add(node.right);
        }

        return res;
    }
}
